include/mata/nfa.hh

```diff
@@ -282,55 +301,185 @@ public:
+    {
+        clear_initial();
+        for (const State& st: vec) { this->make_initial(st); }
+    }
+
     bool has_initial(const State &state_to_check) const {return initialstates.count(state_to_check);}
+
     void remove_initial(State state)
     {
         assert(has_initial(state));
         this->initialstates.remove(state);
     }
 
-    void make_final(State state) {
+    /**
+     * Clear final states set.
+     */
+    void clear_final() { finalstates.clear(); }
+
+    /**
+     * Make @p state final.
+     * @param state[in] State to be added to final states.
+     */
+    void make_final(const State state)
+    {
         if (this->get_num_of_states() <= state) {
             throw std::runtime_error("Cannot make state final because it is not in automaton");
         }
 
         this->finalstates.insert(state);
     }
+
+    /**
+     * @brief Set final states set with @p state.
+     *
+     * Overwrite the previous final states set.
+     *
+     * @param state[in] State to be set as the new final state.
+     */
+    void set_final(const State state)
+    {
+        clear_final();
+        make_final(state);
+    }
+
+    /**
+     * Make @p vec of states final states.
+     * @param vec[in] Vector of states to be added to final states.
+     */
     void make_final(const std::vector<State>& vec)
-    { // {{{
+    {
         for (const State& st : vec) { this->make_final(st); }
-    } // }}}
+    }
+
+    /**
+     * @brief Set final states set with @p state.
+     *
+     * Overwrite the previous final states set.
+     *
+     * @param vec[in] Vector of states to be set as new final states.
+     */
+    void set_final(const std::vector<State>& vec)
+    {
+        clear_final();
+        for (const State& st: vec) { this->make_final(st); }
+    }
+
     bool has_final(const State &state_to_check) const { return finalstates.count(state_to_check); }
+
     void remove_final(State state)
     {
         assert(has_final(state));
         this->finalstates.remove(state);
     }
 
     /**
-     * @brief Returns a newly created state.
+     * Add a new state to the automaton.
+     * @return The newly created state.
      */
     State add_new_state();
+
     bool is_state(const State &state_to_check) const { return state_to_check < transitionrelation.size(); }
 
+    /**
+     * @brief Get set of reachable states.
+     *
+     * Reachable states are states accessible from any initial state.
+     * @return Set of reachable states.
+     */
+    StateSet get_reachable_states() const;
+
+    /**
+     * @brief Get set of terminating states.
+     *
+     * Terminating states are states leading to any final state.
+     * @return Set of terminating states.
+     */
+    StateSet get_terminating_states() const;
+
+    /**
+     * @brief Get a set of useful states.
+     *
+     * Useful states are reachable and terminating states.
+     * @return Set of useful states.
+     */
+    StateSet get_useful_states();
```

Comment:
Is this well established terminology (useful states)?

Reply:
I cannot come up with a better name. Does anyone have any ideas? Otherwise, I will take a look elsewhere for some suggestions to consider.

Reply:
There are several discussions about useless states (unreachable **or** non-terminating) and useful states (reachable **and** terminating).

I would say that yes, I consider this a somewhat established terminology. Therefore, unless anyone has a better idea, I would keep the name, at least for now.

Reply:
Seeing as nobody around could come up with a better name, I will keep the term `useful` to denote reachable **and** terminating states.

Reply:
Yes, I didn't have issue with the name, just if there is some well-established naming for such states, I would use that :-). But apparently, there is not.